staging/src/k8s.io/apiserver/pkg/server/filters/compression.go

```diff
@@ -0,0 +1,147 @@
+package filters
+
+import (
+	"compress/gzip"
+	"compress/zlib"
+	"errors"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/emicklei/go-restful"
+)
+
+type Compressor interface {
+	io.WriteCloser
+	Flush() error
+}
+
+const (
+	header_AcceptEncoding  = "Accept-Encoding"
+	header_ContentEncoding = "Content-Encoding"
+
+	encoding_gzip    = "gzip"
+	encoding_deflate = "deflate"
+)
+
+func WithCompression(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		//don't compress watches
+		if req.URL.Query().Get("watch") == "true" {
```

Comment:
plumb through a `RequestContextMapper` from https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/installer.go#L188 and then use https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/request/requestinfo.go#L229 and https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/request/requestinfo.go#L35